JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,598 @@
+``` jsx
+// exporting when static members are not supported by the transpiler:
+class ComponentName extends React.Component {
+  // ...
+}
+
+ComponentName.propTypes = {
+  // ..
+};
+
+export {ComponentName};
+```
+
+Exporting functional components:
+``` jsx
+// stateless components with no prop validation
+export function FunctionalComponent() {
+  // ...
+};
+
+
+// stateless components with static prop validation
+function FunctionalComponent(/* props */) {
+  // ...
+}
+
+FunctionalComponent.propTypes = {
+  // ...
+};
+
+export {FunctionalComponent};
+```
+
+### Rendering Arrays of Data
+
+For iterating arrays:
+- use **map** method to render items
+- you should have a unique key for each element
+- **don't use array's index as a key**, unless you have no choice (no unique property in each element of the array)
+
+``` jsx
+render() {
+  // ...
+  return (
+    <ul>
+      {
+        booksArray.map((book) => (
+          <li key={book.id}>
+            <a href={book.href}>{book.title}</a>
+          </li> 
+        )); 
+      }
+    </ul>
+  );
+}
+```
+
+### Binding
+- don't bind functions inline or in a class constructor
+- use [`developit/decko`](https://github.com/developit/decko) lib's *@bind* decorator for binding
+
+``` jsx
+// bad
+export class ComponentName extends React.Component {
+  onClickHandler(e) {/**/}
+
+  render() {
+    return (
+      <ChildComponent onClick={this.onClickHandler.bind(this)} />
+    );
+  }
+}
+
+// bad
+export class ComponentName extends React.Component {
+  constructor(args) {
+    super(args);
+    this.onClickHandler = this.onClickHandler.bind(this);
+  }
+
+  onClickHandler(e) {/**/}
+
+  render() {
+    return (
+      <ChildComponent onClick={this.onClickHandler} />
+    );
+  }
+}
+
+// good
+export class ComponentName extends React.Component {
+
+  @bind
+  onClickHandler(e) {/**/}
+
+  render() {
+    return (
+      <ChildComponent onClick={this.onClickHandler} />
+    );
+  }
+}
+```
+
+### CSS Modules and Components
+Each component should have it's own stylesheet if CSS is needed for it.
+
+```
+...
+├─ Login
+⎮   └─ Login.js
+⎮   └─ Login.scss
+...
+```
+
+Importing:
+
+``` jsx
+// ...
+import styles from './ComponentName.scss';
+
+export class ComponentName extends React.Component {
+  // ...
+  render() {
+    return (
+      <Heading className={styles.specialHeading} />
+    );
+  }
+}
+```
+
+If you need to pass more than one CSS class name to a component or want to use conditional classes, use the [`classnames`](https://www.npmjs.com/package/classnames) library.
+
+``` jsx
+import classNames from 'classnames';
+// ...
+import styles from './ComponentName.scss';
+
+export class ComponentName extends React.Component {
+  // ...
+  render() {
+    return (
+      <Heading
+        className={classNames(styles.largeHeading, {
+          [styles.importantHeading]: isImportant,
+        })}
+      />
+    );
+  }
+}
+```
+
```

Comment:
Maybe add one more item here defining the max number of lines in a single file? @andreicek